### ISSUE TITLE
Custom Ontology

### ISSUE THREAD

**Issue by a1012** [biocypher/biocypher#225]
Hi @slobentanzer
I am tying to build a knowledge graph on custom dataset.
I have created .ttl file and want to build knowledge graph.
Could you please guide me on using own ttl file for building graph.
Sample ontology : 
0 Category Drug ;
    Entities P276-00 ;
    Relationship P276-00 inhibits growth of head and neck cancers .

1 Category Proteins ;
    Entities CDK/CCND1 proteins ;
    Relationship CDK/CCND1 proteins are overexpressed in head and neck cancer .

2 Category Disease ;
    Entities Head and neck cancer .

There is no such example.

**Comment by slobentanzer** [comment]
Hi @a1012,
Thanks for the question. As described in the [Ontology Tutorial](https://biocypher.org/tutorial-ontology.html#hybridising-ontologies), you can point BioCypher to any (valid) OWL file using the YAML configuration. In your case, you need to replace the link to the Biolink ttl with your own file. More should not be required.

**Comment by a1012** [comment]
@slobentanzer 
I  followed your instruction and replaced biolink ttl with my own ttl file in the below code:

`bc = BioCypher(
    biocypher_config_path='/content/sample_data/biocypher_config.yaml',
    offline=True,
)
print(bc)
bc.show_ontology_structure()
`


biocypher_config.yaml:
biocypher:  # biocypher settings

  # Ontology configuration
  head_ontology:
    url: /content/sample_data/NER.ttl
    root_node: entity
  ### Optional parameters ###
  # Logging granularity
  # Set debug to true if more granular logging is desired
  debug: true

  # Set to change the log directory
  log_directory: biocypher-log

  # Set to change the output directory
  output_directory: biocypher-out


But the error is coming : FileNotFoundError: [Errno 2] No such file or directory: 'config/schema_config.yaml'

**Comment by slobentanzer** [comment]
Did you go through the tutorial? You need a schema configuration, which is not found.

**Comment by a1012** [comment]
Yeah I have gone through the tutorials, the examples contain 2 config files, one is schema file and another is biocypher_config but there is no mention how schema file is prepared  for custom dataset.
What i understood is that ttl file can be used instead of schema_config.yaml

**Comment by a1012** [comment]
@slobentanzer I created schema and given ttl file instead of biolink ontology. Could you please tell me where am I going wrong?
I have uploaded schema and ttl (ontology) in txt file for your perusal
[ttl_file.txt](https://github.com/biocypher/biocypher/files/12288792/ttl_file.txt)
[schema.txt](https://github.com/biocypher/biocypher/files/12288746/schema.txt)

The code:
`# Extract id, label, and property dictionary
def node_generator():
      for protein in proteins:
          yield (
              protein.get_id(),
              protein.get_label()
          )
# Create BioCypher driver
bc = BioCypher(
    biocypher_config_path="/content/sample_data/biocypher_config.yaml",
    schema_config_path="/content/sample_data/schema_config.yaml",
)

# Visualise ontology schema
bc.show_ontology_structure()
# # Run the import
bc.write_nodes(node_generator())
bc.write_edges(edge_generator())

# # Write command line call
bc.write_import_call()
`

**Comment by slobentanzer** [comment]
@a1012 can you tell me what your error is?

**Comment by a1012** [comment]
@slobentanzer 
The error is :

  File "/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py", line 3553, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)

  File "<ipython-input-10-64458d54a7fc>", line 31, in <cell line: 31>
    bc.show_ontology_structure()

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_core.py", line 490, in show_ontology_structure
    self._get_ontology()

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_core.py", line 203, in _get_ontology
    self._ontology = Ontology(

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_ontology.py", line 281, in __init__
    self._main()

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_ontology.py", line 290, in _main
    self._load_ontologies()

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_ontology.py", line 314, in _load_ontologies
    self._head_ontology = OntologyAdapter(

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_ontology.py", line 90, in __init__
    self._rdf_graph = self._load_rdf_graph(ontology_file)

  File "/usr/local/lib/python3.10/dist-packages/biocypher/_ontology.py", line 195, in _load_rdf_graph
    g.parse(ontology_file, format=self._get_format(ontology_file))

  File "/usr/local/lib/python3.10/dist-packages/rdflib/graph.py", line 1503, in parse
    raise se

  File "/usr/local/lib/python3.10/dist-packages/rdflib/graph.py", line 1494, in parse
    parser.parse(source, self, **args)

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 2021, in parse
    p.loadStream(stream)

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 479, in loadStream
    return self.loadBuf(stream.read())  # Not ideal

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 485, in loadBuf
    self.feed(buf)

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 511, in feed
    i = self.directiveOrStatement(s, j)

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 532, in directiveOrStatement
    return self.checkDot(argstr, j)

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 1214, in checkDot
    self.BadSyntax(argstr, j, "expected '.' or '}' or ']' at end of statement")

  File "/usr/local/lib/python3.10/dist-packages/rdflib/plugins/parsers/notation3.py", line 1730, in BadSyntax
    raise BadSyntax(self._thisDoc, self.lines, argstr, i, msg)

  File "<string>", line unknown
BadSyntax

**Comment by a1012** [comment]
Hi @slobentanzer 
I want your guidance on my usecase to create own custom ontology using BioCypher

**Comment by slobentanzer** [comment]
Hi @a1012,

> I have article on biochemical industry which is in pdf format and want to generate ontology out of it.

I am not sure I understand what you mean by "ontology". Usually, ontologies used in BioCypher graphs are highly curated domain expertise, result of many experts agreeing on the representation of knowledge in a particular field. Popular examples are mentioned in the [BioCypher docs on ontology](https://biocypher.org/tutorial-ontology.html). We don't generate ontologies from single articles, and I would not recommend approaching building an ontology if you are not a domain expert. As such, I don't understand the task you envision.

> If I am able to extract entity-relationship then will biocypher be able to categories entities

BioCypher is a knowledge representation framework. If you managed to extract entities from literature, the categorisation should already have happened. What you can do with BioCypher is then build a knowledge graph that maps these entities to one or more domain-specific ontologies (that already exist). At a general level, you probably want to use something like Biolink as an ontology.

I hope this clarifies things.

**Comment by a1012** [comment]
Hi @slobentanzer 
Can we merge two ontologies  using BioCypher? For example: [Experimental Factor Ontology, NCBO Bio Portal]
If yes how we can do it? Which format do we have to give it to BioCypher like it is mentioned that in tutorial some ETL adapter we have write in order to give data in proper format accepted by BioCypher.
Also, If possible could we please keep a session on BioCypher for my organization and team for them to understand what exactly is BioCypher?

**Comment by slobentanzer** [comment]
Hi @a1012,
thanks for your continued interest. What you are asking is described in the documentation, please try to follow that first. GitHub issues are meant to be used for actual problems that you cannot solve, such as bugs in the software or missing features. They are not a teaching channel, and not for free consulting. Thanks for your understanding.

**Comment by jdreo** [comment]
For anyone coming here and having a similar question, I believe the misunderstanding comes from the fact that for many people in the semantic web community, an ontology can be a file handling both the *vocabulary* (i.e. the types hierarchy that BioCypher uses) *and* the *instances* (i.e. the nodes and edges, with their IDs). So "building up an ontology with BioCypher" makes sense if you use the RDF backend and produce a file containing the instances.

**Comment by slobentanzer** [comment]
@jdreo thanks for the clarification; closing this as stale. For any informal questions or help requests, please rather ask in our community forum at https://biocypher.zulipchat.com and leave the GitHub issues for technical issues. :)
